llms/mlx_lm/models/gemma3_text.py

```diff
@@ -0,0 +1,249 @@
+                mask = mask[..., -key_len:]
+
+        output = mx.fast.scaled_dot_product_attention(
+            queries, keys, values, scale=self.scale, mask=mask
+        )
+        output = output.transpose(0, 2, 1, 3).reshape(B, L, -1)
+        return self.o_proj(output)
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, dims: int, eps: float = 1e-5):
+        super().__init__()
+        self.weight = mx.ones((dims,))
+        self.eps = eps
+
+    def __call__(self, x):
+        return mx.fast.rms_norm(x, 1.0 + self.weight, self.eps)
+
+
+class MLP(nn.Module):
+    def __init__(self, dim, hidden_dim):
+        super().__init__()
+        self.gate_proj = nn.Linear(dim, hidden_dim, bias=False)
+        self.down_proj = nn.Linear(hidden_dim, dim, bias=False)
+        self.up_proj = nn.Linear(dim, hidden_dim, bias=False)
+
+    def __call__(self, x) -> mx.array:
+        # This should not be GELU approx, jax.nn.gelu
+        return self.down_proj(nn.gelu_approx(self.gate_proj(x)) * self.up_proj(x))
+
+
+class TransformerBlock(nn.Module):
+    def __init__(self, args: ModelArgs, layer_idx: int):
+        super().__init__()
+        self.num_attention_heads = args.num_attention_heads
+        self.hidden_size = args.hidden_size
+        self.self_attn = Attention(args, layer_idx)
+        self.mlp = MLP(args.hidden_size, args.intermediate_size)
+        self.input_layernorm = RMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+        self.post_attention_layernorm = RMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+        self.pre_feedforward_layernorm = RMSNorm(
+            args.hidden_size, eps=args.rms_norm_eps
+        )
+        self.post_feedforward_layernorm = RMSNorm(
+            args.hidden_size, eps=args.rms_norm_eps
+        )
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Any] = None,
+    ) -> mx.array:
+        r = self.self_attn(self.input_layernorm(x), mask, cache)
+        h = x + self.post_attention_layernorm(r)
+        r = self.mlp(self.pre_feedforward_layernorm(h))
+        out = h + self.post_feedforward_layernorm(r)
+        return out
+
+
+class Gemma3Model(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+        self.vocab_size = args.vocab_size
+        self.num_hidden_layers = args.num_hidden_layers
+        assert self.vocab_size > 0
+        self.embed_tokens = nn.Embedding(args.vocab_size, args.hidden_size)
+        self.layers = [
+            TransformerBlock(args=args, layer_idx=layer_idx)
+            for layer_idx in range(args.num_hidden_layers)
+        ]
+        self.norm = RMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        mask: mx.array = None,
+        cache=None,
+    ):
+
+        h = self.embed_tokens(inputs)
+        h *= self.args.hidden_size**0.5  # persistent precision issue in scaling
```

Comment:
Can you elaborate on the precision comment here?

Reply:
Compared to the transformer implementation I was able to get up to `1e-2` differences pre and post embedding scaling. Could not determine what type promotion wizardry was causing it. Was also mentioned by @zcbenz for gemma2: https://github.com/ml-explore/mlx-examples/issues/1057

Reply:
There is no type promotion wizardry there. The `h` has a dtype of probably float16 or bfloat16 and multiplying it by a the Python scalar will not change it's type.

When you increase the magnitude (which they are here by quite a bit actually) it's sort of expected that the numerical differences between the two implementations go up as well.. but if you check on a relative basis (e.g. `(x - y).abs().max() / y.max()` it should not look so bad.

Reply:
Sorry, this is an artifact from VLM. It was related to RMSNorm and was fixed.

The only issue I noticed was that long context doesnt work well. The model cant generate coherent asnwers for 10K.

Reply:
Interesting 🤔, maybe there is an issue with the sliding window attention.. we'll have to debug that but we can still land this in the meantime.

Reply:
When I checked the relative basis it was fine.

Reply:
scaling_factor is in `float32` in transformers, which in torch will promote `h` AFAIK. something is afoot didn't have time to get to the bottom of it.

Reply:
It doesn't manually cast but won't it implicitly cast? given that `embed_scale` is a python `float`.

Your solution looks good!

Reply:
> It doesn't manually cast but won't it implicitly cast? given that embed_scale is a python float.

No it doesn't up cast (the output as the same type as the input). Whether or not it implicitly up casts and down casts inside the kernel depends on the platform and what precision it supports natively vs what is emulated.

```
>>> torch.tensor([1.0, 2.0]).to(torch.float16) * 1.0
tensor([1., 2.], dtype=torch.float16)
>>> 
```